pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -218,50 +218,87 @@ func (p *DNSProxy) checkRestored(endpointID uint64, destPort uint16, destIP stri
 	return false
 }
 
+// skipIPInRestoration skips IPs that are allowed but have never been used,
+// but only if at least one server has been used so far.
+func (p *DNSProxy) skipIPInRestoration(ip string) bool {
+	p.RLock()
+	defer p.RUnlock()
```

Comment:
nit: Depending on how frequently Cilium hits `GetRules()`, we could amortize locking costs by assembling the IPs in a separate loop then grabbing the lock again later on. But I'm fine with this solution unless we get specific indicators that it will be a performance problem :+1:

Reply:
I did hoist the RLock one loop up, c.f. https://github.com/cilium/cilium/pull/24672#discussion_r1159482992

The outer loop could probably be fixed as well via your suggestion, but I'll leave that as a follow-up item for now.